### PR TITLE
Feature/disable accepted story fields

### PR DESCRIPTION
**Pull request by jprask**
Disable input fields for read only stories [#23438](https://central.cm42.io/projects/central-board-v2#story-23438).

Before:
before
After:
after

Note that if a component has no content and the story is read-only the render method will return `null` (i.e. "Attachments" on the example above)

### CODE REVIEW

File: app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryAttachments.js [hunk 1]
```diff
@@ -58,40 +58,45 @@ class ExpandedStoryAttachments extends React.Component {
...
-                  {!isDragReject ?
-                    I18n.t('upload_new_file') :
-                    I18n.t('reject_new_file')}
-                </div>
-              </div>
-            )
-          }}
-        </Dropzone>
+        {
+          canEdit(story) &&
```
Comment (talyssonoc) [inline app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryAttachments.js]: We don't need this `canEdit(story)` here because we already return `null` if it can't edit in line 61 😄

File: app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryDescription.js [hunk 2]
```diff
@@ -36,19 +36,22 @@ class ExpandedStoryDescription extends React.Component {
   };
 
   descriptionTextArea(description) {
-    const { onEdit } = this.props;
+    const { onEdit, readOnly } = this.props;
```
Comment (talyssonoc) [inline app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryDescription.js]: Why is it called `readOnly` for his one instead of `disabled`?

File: app/assets/javascripts/components/story/ExpandedStory/index.js [hunk 3]
```diff
@@ -55,12 +55,14 @@ export class ExpandedStory extends React.Component {
               story={story}
               titleRef={this.titleRef}
               onEdit={(newAttributes) => editStory(story.id, newAttributes)}
+              disabled={!Story.canEdit(story)}
```
Comment (talyssonoc) [inline app/assets/javascripts/components/story/ExpandedStory/index.js]: We should put this value in a variable and re-use it here and on line 65.

File: app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryDescription.js (hunk 2 above)
Comment (jprask) [reply app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryDescription.js]: It uses the readonly attribute for the input field so the user can copy the text. We could use disabled as well if that is not really necessary.

Comment (talyssonoc) [reply app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryDescription.js]: We can have the component to be called `disabled` but pass it as the value of `readOnly` of the `textarea`:

```js
<textarea
  readOnly={ disabled }
```

File: app/assets/javascripts/components/story/ExpandedStory/index.js (hunk 3 above)
Comment (jprask) [reply app/assets/javascripts/components/story/ExpandedStory/index.js]: Good idea.

File: spec/javascripts/components/story/expanded_story/expanded_story_labels_spec.js [hunk 4]
```diff
@@ -3,39 +3,88 @@ import { shallow, mount } from 'enzyme';
...
+      wrapper.instance().handleDelete(0);
+
+      expect(onRemoveLabel).not.toHaveBeenCalled();
+    });
+
+    describe('when story has no labels', () => {
+      it('renders nothing', () => {
+        const { wrapper } = setup({ disabled: true }, []);
+
+        expect(wrapper.type()).toBeFalsy();
```
Comment (talyssonoc) [inline spec/javascripts/components/story/expanded_story/expanded_story_labels_spec.js]: What is this actually testing?

File: spec/javascripts/components/story/expanded_story/expanded_story_notes_spec.js [hunk 5]
```diff
@@ -51,47 +53,89 @@ describe('<ExpandedStoryNotes />', () => {
...
+          story={story}
+          onCreate={onCreate}
+          onDelete={onDelete}
+          disabled={true}
+        />
+      );
+      
+      const btnExsists = wrapper.exists('.create-note-button');
+      const formExists = wrapper.exists('.create-note-text');
+      expect(btnExsists && formExists).toBe(false);
```
Comment (talyssonoc) [inline spec/javascripts/components/story/expanded_story/expanded_story_notes_spec.js]: We should avoid `&&` in assertions like that. If we need to check the presence of two things, we should create to `it`.

File: spec/javascripts/components/story/expanded_story/expanded_story_task_spec.js [hunk 6]
```diff
@@ -32,32 +33,66 @@ describe('<ExpandedStoryTask />', () => {
...
+    });
 
-  describe('onHandleSubmit', () => {
-    it('calls onSave with a task', () => {
-      const task = 'New Task';
-      const onSaveSpy = sinon.spy();
-      const { wrapperInstance } = setup({ onSave: onSaveSpy });
-      wrapperInstance.setState({ task })
+    it('disables the add task button if text area is empty', () => {
+      const { input } = setup();
```
Comment (talyssonoc) [inline spec/javascripts/components/story/expanded_story/expanded_story_task_spec.js]: Do we need two setups? If no, we can just destructure both from the same call:

```suggestion
      const { input, button } = setup();
```

File: spec/javascripts/components/story/expanded_story/expanded_story_task_spec.js [hunk 7]
```diff
@@ -32,32 +33,66 @@ describe('<ExpandedStoryTask />', () => {
...
+      input.simulate('change', { target: { value: '' } });
+      expect(button.prop('disabled')).toBe(true);
+    });
+  
+    describe('onHandleSubmit', () => {
+      it('calls onSave with a task', () => {
+        const task = 'New Task';
+        const onSaveSpy = sinon.spy();
+        const { wrapperInstance } = setup({ onSave: onSaveSpy });
+        wrapperInstance.setState({ task })
```
Comment (talyssonoc) [inline spec/javascripts/components/story/expanded_story/expanded_story_task_spec.js]: Do we need to set the state like this? It's not a good practice to set the state of a component externally.

File: spec/javascripts/components/story/expanded_story/expanded_story_description_spec.js [hunk 8]
```diff
@@ -5,9 +5,27 @@ import ExpandedStoryDescription from 'components/story/ExpandedStory/ExpandedSto
 describe('<ExpandedStoryDescription />', () => {
   const defaultProps = () => ({
     story: {},
-    onEdit: sinon.spy()
+    onEdit: sinon.spy(),
+    disabled: false
   });
 
+  const mountTextarea = propOverrides => {
```
[the remark this answers is not in the document]
Comment (jprask) [reply spec/javascripts/components/story/expanded_story/expanded_story_description_spec.js]: Will do.

File: spec/javascripts/components/story/expanded_story/expanded_story_labels_spec.js (hunk 4 above)
Comment (jprask) [reply spec/javascripts/components/story/expanded_story/expanded_story_labels_spec.js]: We need to test if the render method returns null, we can also use `expect(wrapper.htlm()).toBeNull()` which is more verbose.

File: spec/javascripts/components/story/expanded_story/expanded_story_notes_spec.js (hunk 5 above)
Comment (jprask) [reply spec/javascripts/components/story/expanded_story/expanded_story_notes_spec.js]: Got it.

File: spec/javascripts/components/story/expanded_story/expanded_story_task_spec.js (hunk 6 above)
Comment (jprask) [reply spec/javascripts/components/story/expanded_story/expanded_story_task_spec.js]: I will check it.

File: spec/javascripts/components/story/expanded_story/expanded_story_task_spec.js (hunk 7 above)
Comment (jprask) [reply spec/javascripts/components/story/expanded_story/expanded_story_task_spec.js]: We could simulate some events to get there, i'm about to push these adjustments.

File: app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryOwnedBy.js [hunk 9]
```diff
@@ -1,24 +1,26 @@
 import React from 'react';
 import PropTypes from 'prop-types';
 import SelectUser from '../select_user/SelectUser';
-import { editingStoryPropTypesShape } from '../../../models/beta/story';
+import { canEdit, editingStoryPropTypesShape } from '../../../models/beta/story';
```
Comment (talyssonoc) [inline app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryOwnedBy.js]: Looks like we're not using `canEdit` here?

File: app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryTask.js [hunk 10]
```diff
@@ -35,7 +35,9 @@ class ExpandedStoryTask extends Component {
   }
 
   render() {
-    const { story, onToggle, onDelete } = this.props;
+    const { story, onToggle, onDelete, disabled } = this.props;
+
+    if (!story.tasks.length && disabled) return null;
```
Comment (talyssonoc) [inline app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryTask.js]: Can we check disabled first and then the length of the tasks as we do for the other components? Just to keep it standard 😄

File: app/assets/javascripts/components/story/task/Task.js [hunk 11]
```diff
@@ -2,30 +2,35 @@ import React from 'react';
...
     </label>
-    <span
-      title={I18n.t('delete')}
-      className='delete-btn'
-      onClick={onDelete}
-    >
-      { I18n.t('delete') }
-    </span>
+    {
+      !disabled &&
```
Comment (talyssonoc) [inline app/assets/javascripts/components/story/task/Task.js]: Can we add parentheses around the span here? Just to make it easier to read.

File: spec/javascripts/components/story/expanded_story/expanded_story_attachments_spec.js [hunk 12]
```diff
@@ -40,11 +41,65 @@ describe('<ExpandedStoryAttachments />', () => {
...
+      it('renders nothing', () => {
+        const documents = [];
+        const story = { documents, _editing: { documents } };
+
+        const wrapper = mount(
+          <ExpandedStoryAttachments
+            story={story}
+            onAdd={onAdd}
+            onDelete={onDelete}
+            disabled={true}
```
Comment (talyssonoc) [inline spec/javascripts/components/story/expanded_story/expanded_story_attachments_spec.js]: Can we use `false` here to show that it won't render when there are no attachments even if it's enabled?

File: spec/javascripts/components/story/note/note_spec.js [hunk 13]
```diff
@@ -10,22 +10,46 @@ describe('<Note/>', () => {
...
-      const wrapper = shallow(
-        <Note
-          note={note}
-          onDelete={onDeleteSpy}
-        />
-      );
+    const button = wrapper.find('.delete-note-button');
 
-      const button = wrapper.find('.delete-note-button');
+    return { wrapper, button, onDeleteSpy };
```
Comment (talyssonoc) [inline spec/javascripts/components/story/note/note_spec.js]: can we rename `button` to `deleteButton` to make the tests that use it easier to understand? I got a little confused when I saw this test https://github.com/Codeminer42/cm42-central/pull/540/files#diff-5b2bbd29d2e9844431270fb14ac57c98R31

File: spec/javascripts/components/story/task/task_spec.js [hunk 14]
```diff
@@ -45,4 +46,24 @@ describe('<Task />', () => {
       expect(onToggleCheckedBoxSpy).toHaveBeenCalled();
     });
   });
+
+  describe('when task is read-only', () => {
+    it('does not render a Delete span', () => {
+      const { span } = setup({ disabled: true });
```
Comment (talyssonoc) [inline spec/javascripts/components/story/task/task_spec.js]: Which span is that? Can we rename it to something clearer? Maybe it could be renamed to `deleteLink`

File: app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryOwnedBy.js (hunk 9 above)
Comment (jprask) [reply app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryOwnedBy.js]: True, it was refactored and I forgot about that

File: app/assets/javascripts/components/story/task/Task.js (hunk 11 above)
Comment (jprask) [reply app/assets/javascripts/components/story/task/Task.js]: Yeah, there are some other files with similar code, looking into it rn

File: spec/javascripts/components/story/expanded_story/expanded_story_attachments_spec.js (hunk 12 above)
Comment (jprask) [reply spec/javascripts/components/story/expanded_story/expanded_story_attachments_spec.js]: I think we can't do that because if its enabled the story is not read-only, so it should render the attachments section and the upload button, right ?

File: spec/javascripts/components/story/note/note_spec.js (hunk 13 above)
Comment (jprask) [reply spec/javascripts/components/story/note/note_spec.js]: Yeah that's nice

File: spec/javascripts/components/story/expanded_story/expanded_story_attachments_spec.js (hunk 12 above)
Comment (talyssonoc) [reply spec/javascripts/components/story/expanded_story/expanded_story_attachments_spec.js]: Can we change the description of the context (https://github.com/Codeminer42/cm42-central/pull/540/files/a17c6c997a68ee473a7678004bc95029d26abd58#diff-92c66b10be3e6fd942203618c2216a8dR87) then to `'when it is disabled and there are no attachments'`? It'll make it clearer when this condition happens.

Comment (talyssonoc) [reply spec/javascripts/components/story/expanded_story/expanded_story_attachments_spec.js]: nevermind, it's right!
